File: Python/0213-m-house-robber-ii.py
"""
Dynamic Programming
all cases {
    rob 0 -> max == 7
    rob 4(last) -> max == 9
} find max
Time: O(n)
Space: O(1)
"""
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]

        return max(self.subRob(nums, 0, len(nums) - 2), self.subRob(nums, 1, len(nums) - 1))

    def subRob(self, nums, start, end):
        prev = 0
        cur = nums[start]

        for i in range(start + 1, end + 1):
            next = max(cur, prev + nums[i])
            prev = cur
            cur = next
        
        return cur

File: leetcode/C++/0213-m-house-robber-ii.cpp
// DP
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        
        vector<int> A(n-1, 0);
        int res = INT_MIN;
        for (int i = 0; i < n - 1; ++i)
            A[i] = nums[i];
        res = max(res, helper(A));
        
        for (int i = 0; i < n - 1; ++i) 
            A[i] = nums[i + 1];
        res = max(res, helper(A));
        
        return res;
    }

    int helper(vector<int>& nums) {
        int prev_max = 0;
        int curr_max = 0;
        for (int x : nums) {
            int tmp = curr_max;
            curr_max = max(prev_max + x, curr_max);
            prev_max = tmp;
        }
        return curr_max;
    }
};